api/src/main/java/com/spotify/ffwd/util/HighFrequencyDetector.java

```diff
@@ -0,0 +1,157 @@
+import com.spotify.ffwd.model.Metric;
+import com.spotify.ffwd.statistics.HighFrequencyDetectorStatistics;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class HighFrequencyDetector {
+
+  /** Allow to drop high frequency metrics. */
+  @Inject
+  @Named("dropHighFrequencyMetric")
+  Boolean dropHighFrequencyMetric;
+
+  /** Minimum number of milliseconds allowed between data points. */
+  @Inject
+  @Named("minFrequencyMillisAllowed")
+  Integer minFrequencyMillisAllowed;
+
+  /** Minimum number of times high frequency detected before metrics are dropped. */
+  @Inject
+  @Named("minNumberOfTriggers")
+  Integer minNumberOfTriggers;
+
+  /** Milliseconds high frequency triggers are refreshed. */
+  @Inject
+  @Named("highFrequencyDataRecycleMS")
+  Long highFrequencyDataRecycleMS;
+
+  @Inject
+  private HighFrequencyDetectorStatistics statistics;
+
+  /** High frequency metrics counter. */
+  // TODO review to use ConcurrentHashMap vs AtomicReference
+  final AtomicReference<Map<Integer, Integer>> highFrequencyMetrics =
+      new AtomicReference<>(new HashMap<>());
+
+  final AtomicLong highFrequencyTriggersTS;
+  private AtomicBoolean highFrequencyRecycleLock;
+
+  /* TODO
+      1. unit tests
+          a. sendBatch with many points
+          b. send both with many points
+          c. check swap of high freq hashmap
+   */
+
+  @Inject
+  public HighFrequencyDetector() {
+    this.highFrequencyTriggersTS = new AtomicLong(System.currentTimeMillis());
+    this.highFrequencyRecycleLock = new AtomicBoolean(false);
+  }
+
+  public List<Metric> detectHighFrequencyMetrics(final List<Metric> metrics) {
+    List<Metric> newList = new ArrayList<>();
+
+    // Groups metrics by metric hash code and finds timestamp deltas of ordered data points.
+    // {hashcode -> -1, hashcode1 -> 10}
+    Map<Integer, Integer> groupedMetrics =
+        metrics.stream()
+            .sorted(Comparator.comparing(Metric::getTime))
+            .collect(
+                Collectors.groupingBy(
+                    Metric::hashCode,
+                    Collectors.collectingAndThen(
+                        Collectors.toList(),
+                        list -> {
+                          int size = list.size();
+                          return IntStream.range(1, size)
+                              .map(
+                                  x ->
+                                      (int)
+                                          (list.get(size - x).getTime().getTime()
+                                              - list.get(size - x - 1).getTime().getTime()))
+                              .filter(d -> (d >= 0 && d < minFrequencyMillisAllowed))
```

Comment:
perhaps extract this predicate and the map function above it (the part where two Metric instances have their timestamps subtracted) into private methods to lower the amount of indentation / operations here

Reply:
Yep, moved.